storage/google/cloud/storage/blob.py

```diff
@@ -2028,12 +2028,16 @@ def upload_parallel(
 
     def _upload_from_list(self, content_type, client, predefined_acl):
         while self._total_files > self._file_count:
-            self.upload_from_filename(
-                self._files_list[self._file_count], content_type, client, predefined_acl
-            )
             self._lock.acquire()
+            current_file_index = self._file_count
             self._file_count += 1
             self._lock.release()
+            self.upload_from_filename(
+                self._files_list[current_file_index],
+                content_type,
+                client,
+                predefined_acl,
+            )
```

Comment:
Now there is a different kind of a problem. If the upload fails, the counter is still increased. There should be some sort of a retry mechanism to recover from such state, or an index tracking, to handle each file individually.

Reply:
I think in the Storage module they didn't apply mechanism of retry, that's why they have a  different task in git  to implement retry mechanism in storage. 
 Please refer issue [7907].

Reply:
Regardless of that, the possibility of racing should be eliminated.